usr.bin/mail/mail.1
.\"	$NetBSD$
.\"
.\" Copyright (c) 1980, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)mail.1	8.8 (Berkeley) 4/28/95
.\"
.Dd December 15, 2014
.Dt MAIL 1
.Os
.Sh NAME
.Nm mail ,
.Nm mailx ,
.Nm Mail
.Nd send and receive mail
.Sh SYNOPSIS
.Nm
.Op Fl EIinv
.Op Fl a Ar file
.Op Fl b Ar bcc-addr
.Op Fl c Ar cc-addr
.Op Fl r Ar rcfile
.Op Fl s Ar subject
.Ar to-addr ...
.Op - Ar sendmail-flags
.Nm
.Op Fl EIiNnv
.Op Fl H Ns Op colon-modifier
.Fl f
.Op Ar name
.Nm
.Op Fl EIiNnv
.Op Fl H Ns Op colon-modifier
.Op Fl u Ar user
.Sh DESCRIPTION
.Nm
is an intelligent mail processing system, which has
a command syntax reminiscent of
.Xr \&ed 1
with lines replaced by messages.
.Pp
.Bl -tag -width flag
.It Fl a
Attach
.Ar file
to the message.
.It Fl b
Send blind carbon copies to
.Ar list .
List should be a comma-separated list of names.
.It Fl c
Send carbon copies to
.Ar list
of users.
.It Fl E
Don't send messages with an empty body.
This is useful for piping errors from cron scripts.
.It Fl f
Read in the contents of your
.Ar mbox
(or the specified file)
for processing; when you
.Ic quit ,
.Nm
writes undeleted messages back to this file.
.It Fl H
Print the header summaries and exit.
The optional colon-modifier string must begin with a
.Ql \&:
and be followed by one or more of the characters described in the
.Sx Specifying messages
section below.
E.g.,
.Do
.Nm
.Fl H Ns :n
.Dc
will display just new message headers.
.It Fl I
Forces mail to run in interactive mode even when
input isn't a terminal.
In particular, the
.Ic \&~
special
character when sending mail is only active in interactive mode.
.It Fl i
Ignore tty interrupt signals.
This is particularly useful when using
.Nm
on noisy phone lines.
.It Fl N
Inhibits the initial display of message headers
when reading mail or editing a mail folder.
.It Fl n
Inhibits reading
.Pa /etc/mail.rc
upon startup.
.It Fl r
Specify an alternate user rcfile to load.
This overrides the value specified in the environment variable
.Ev MAILRC
which in turn overrides the default
.Pa ~/.mailrc
file.
.It Fl s
Specify subject on command line
(only the first argument after the
.Fl s
flag is used as a subject; be careful to quote subjects
containing spaces.)
.It Fl u
Is equivalent to:
.Pp
.Dl mail -f /var/mail/user
.It Fl v
Verbose mode.
The details of delivery are displayed on the user's terminal.
.El
.Ss Sending mail
To send a message to one or more people,
.Nm
can be invoked with arguments which are the names of people to
whom the mail will be sent.
You are then expected to type in your message, followed by a
.Ql control-D
at the beginning of a line.
.Pp
Any flags following the list of recipients, will be passed, together
with their arguments, directly to
.Xr sendmail 1 .
For example to change your
.Dv From
address to
.Dv somebody@somewhere.net
you can specify:
.Pp
.Dl mail recipient -f somebody@somewhere.net
.Pp
To prevent multiple copies of a message being sent to the same
address, duplicate addresses (after alias expansion) are removed from
the
.Ar bcc-addr ,
.Ar cc-addr ,
and
.Ar to-addr
lists.
In addition, addresses on the
.Ar cc-addr
and
.Ar to-addr
lists are removed if they occur on the
.Ar bcc-addr
list and addresses on the
.Ar cc-addr
list are removed if they occur on the
.Ar to-addr
list.
If the
.Ar to-addr
list is empty after these deletions, most systems will insert the line
.Dq To: undisclosed recipients:; .
.Pp
The section below
.Sx Replying to or originating mail ,
describes some features of
.Nm
available to help you compose your letter.
.Ss Reading mail
In normal usage
.Nm
is given no arguments and checks your mail out of the post office,
then prints out a one line header of each message found.
The current message is initially the first message (numbered 1)
and can be printed using the
.Ic print
command (which can be abbreviated
.Ic p ) .
You can move among the messages much as you move between lines in
.Xr \&ed 1 ,
with the commands
.Ic \&+
and
.Ic \&\-
moving backwards and forwards, and simple numbers.
.Ss Disposing of mail
After examining a message you can
.Ic delete
.Pq Ic d
the message or
.Ic reply
.Pq Ic r
to it.
Deletion causes the
.Nm
program to forget about the message.
This is not irreversible; the message can be
.Ic undeleted
.Pq Ic u
by giving its number, or the
.Nm
session can be aborted by giving the
.Ic exit
.Pq Ic x
command.
Deleted messages will, however, usually disappear never to be seen again.
.Ss Specifying messages
Many commands (e.g.,
.Ic delete ,
.Ic from ,
and
.Ic print )
accept a list of messages as an argument.
Messages may be specified by their message number, by a range of
messages, or by a pattern string matching certain fields in the header
as described below.
These message
.Dq specs
may be combined by the usual binary boolean operations
.Ql \*[Am] ,
.Ql \&| ,
and
.Ql \&^ ,
which denote, respectively, a logical
.Dq and ,
.Dq or ,
and
.Dq xor .
Logical expressions may be grouped with parentheses
.Ql \&(
and
.Ql \&)
and
negated with
.Ql \&! .
If the binary operator is missing between two message specs, it is
assumed to be a
.Ql \&| .
This is for simplicity, backwards compatibility, and also to to
facilitate using the
.Ql \&|
symbol to denote a pipe.
(See
.Ar enable-pipes . )
.Pp
Besides the obvious (base10) message numbers, the characters
.Ql \&^ ,
.Ql \&- ,
.Ql \&. ,
.Ql \&+ ,
and
.Ql \&$
denote, respectively, the first message, the message before the
.Dq dot
.Pq the current message ,
the
.Dq dot
message, the message following the
.Dq dot ,
and the last message.
.Pp
A
.Dq message range
consists of two message numbers separated by a
.Ql \&- .
A
.Ql \&*
denotes all messages and is equivalent to
.Ql ^-$ .
.Pp
A pattern is a string (not beginning with any of the above special
characters).
If it does not begin with a
.Ql \&/ ,
it is compared with the senders address.
If it begins with a
.Ql \&/ ,
and
.Ar searchheaders
is not defined, the remainder of the string is compared with the
subject field.
(See
.Ar searchheaders
for searching other header fields or the message body.)
If
.Ar regex-search
is not defined, then the comparison is a simple case insensitive
substring match.
(See
.Ar regex-search
for regular expression matches.)
.Pp
A list of messages may be restricted by a
.Dq colon-modifier
string, i.e., a
.Ql \&:
followed by one or more of the characters:
.Bd -literal -offset indent
d	deleted
e	edited
m	mboxed
n	new
o	old
p	preserved
r	read
s	saved
t	tagged
u	unread and not new
!	invert the meaning of the colon-modifiers
.Ed
.Pp
If there are no address specifications other than colon-modifiers,
the colon-modifiers apply to all messages.
Thus
.Dq Li from netbsd :n
would display the headers of all new messages with
.Ql netbsd
in the sender's address, while
.Dq Li from :!r
and
.Dq Li from :nu
would both display all new and unread messages.
Multiple colon-modifiers may be specified and a single
.Ql \&:
with no letters following indicates the colon-modifier from the
preceding command.
.Pp
For example:
.Bd -literal -offset indent
from 1 12 3-5
.Ed
.Pp
would display the headers from messages 1, 3, 4, 5, and 12.
.Bd -literal -offset indent
from anon \*[Am] ( /foo | /bar )
.Ed
.Pp
would display all headers that had
.Ql anon
in the sender's address and either
.Ql foo
or
.Ql bar
in the subject line.
.Pp
Generally, commands cannot select messages that are not displayed,
such as deleted or hidden messages, the exception being the
.Ic undelete
command.
.Ss Replying to or originating mail
You can use the
.Ic reply
command to set up a response to a message, sending it back to the
person who it was from.
Text you then type in, up to an end-of-file,
defines the contents of the message.
While you are composing a message,
.Nm
treats lines beginning with the character
.Ic \&~
specially.
For instance, typing
.Ic \&~m
(alone on a line) will place a copy
of the current message into the response right shifting it by a tab stop
(see
.Em indentprefix
variable, below).
Other escapes will set up subject fields, add and delete recipients
to the message, and allow you to escape to an editor to revise the
message or to a shell to run some commands.
(These options are given in the summary below.)
.Ss Ending a mail processing session
You can end a
.Nm
session with the
.Ic quit
.Pq Ic q
command.
Messages which have been examined go to your
.Ar mbox
file unless they have been deleted in which case they are discarded.
Unexamined messages go back to the post office.
(See the
.Fl f
option above).
.Ss Personal and system wide distribution lists
It is also possible to create a personal distribution lists so that,
for instance, you can send mail to
.Dq Li cohorts
and have it go
to a group of people.
Such lists can be defined by placing a line like
.Pp
.Dl alias cohorts bill ozalp jkf mark kridle@ucbcory
.Pp
in the file
.Pa \&.mailrc
in your home directory.
The current list of such aliases can be displayed with the
.Ic alias
command in
.Nm .
System wide distribution lists can be created by editing
.Pa /etc/mail/aliases ,
see
.Xr aliases 5
and
.Xr sendmail 1 ;
these are kept in a different syntax.
In mail you send, personal aliases will be expanded in mail sent
to others so that they will be able to
.Ic reply
to the recipients.
System wide
.Ic aliases
are not expanded when the mail is sent,
but any reply returned to the machine will have the system wide
alias expanded as all mail goes through
.Xr sendmail 1 .
.Ss Network mail (ARPA, UUCP, Berknet)
See
.Xr mailaddr 7
for a description of network addresses.
.Pp
.Nm
has a number of options which can be set in the
.Pa .mailrc
file to alter its behavior; thus
.Dq Li set askcc
enables the
.Ar askcc
feature.
(These options are summarized below.)
.Sh SUMMARY
(Adapted from the
.Dq Mail Reference Manual )
.Pp
Each command is typed on a line by itself, and may take arguments
following the command word.
The command need not be typed in its
entirety \- the first command which matches the typed prefix is used.
For commands which take message lists as arguments, if no message
list is given, then the next message forward which satisfies the
command's requirements is used.
If there are no messages forward of the current message, the search
proceeds backwards, and if there are no good messages at all,
.Nm
types
.Dq Li \&No applicable messages
and aborts the command.
.Bl -tag -width delete
.It Ic \&!
Executes the shell
(see
.Xr sh 1
and
.Xr csh 1 )
command which follows.
.It Ic \&\-
Print out the preceding message.
If given a numeric
argument
.Ar n ,
goes to the
.Ar n Ap th
previous message and prints it.
.It Ic \&=
With no argument, it displays the current message number.
Otherwise, set the current message number to its first argument.
.It Ic \&?
Prints a brief summary of commands.
.It Ic \&|
Pipe the current message body through the shell
(see
.Xr sh 1
and
.Xr csh 1 )
command which follows.
.It Ic Detach
Like
.Ic detach
but also saves MIME parts that don't have a filename associated with
them.
For the unnamed parts, a filename is suggested containing the message
and part numbers, and the subtype.
.It Ic Header
.Pq Ic H
Specify or show additional header fields.
This is intended for adding extra header fields like
.Dq Reply-To:
or
.Dq X-Organization:
to the header.
For example:
.Bd -literal -offset 0
  Header X-Mailer: NetBSD mail(1) 9.1
.Ed
.Pp
would add the
.Dq X-Mailer: NetBSD mail(1) 9.1
line to the message header.
Without any arguments, the extra header fields are displayed.
With only a header name (including the
.Sq \&: ) ,
it will delete all extra header fields with that name.
Note: Although some syntax checking is done on the header line, care
should be taken to ensure that it complies with RFC 2821 and 2822.
Also, the extra header lines are not currently displayed by the
.Ic \&~h
tilde command when sending mail (use
.Ic \&~:Header
to see them).
.It Ic More
.Pq Ic M
Like
.Ic more
but also prints out ignored header fields.
.It Ic Page
.Pq Ic \&Pa
A synonym for
.Ic More .
.It Ic Print
.Pq Ic P
Like
.Ic print
but also prints out ignored header fields.
See also
.Ic print ,
.Ic more ,
.Ic page ,
.Ic type ,
.Ic view ,
.Ic ignore ,
and
.Ic retain .
.It Ic Reply
.Pq Ic R
Reply to originator.
Does not reply to other recipients of the original message.
(See
.Ic reply . )
.It Ic Save
.Pq Ic S
Same as
.Ic save
except that all header fields are saved ignoring the
.Ic saveignore
or
.Ic saveretain
lists.
.It Ic Type
.Pq Ic T
Identical to the
.Ic Print
command.
.It Ic View
.Pq Ic V
Like
.Ic Print
but has the opposite MIME decoding behavior.
(See the
.Ar mime-decode-message
variable.)
.It Ic alias
.Pq Ic a
With no arguments, prints out all currently-defined aliases.
With one argument, prints out that alias.
With more than one argument, creates
a new alias or changes an old one.
.It Ic alternates
.Pq Ic alt
The
.Ic alternates
command is useful if you have accounts on several machines.
It can be used to inform
.Nm
that the listed addresses are really you.
When you
.Ic reply
to messages,
.Nm
will not send a copy of the message to any of the addresses
listed on the
.Ic alternates
list.
If the
.Ic alternates
command is given with no argument, the current set of alternative
names is displayed.
.It Ic bounce
Takes a list of messages and prompts for an address to bounce the
messages to.
If no message is specified, the current message is used.
All the original header fields are preserved except for the
.Ql Delivered-To ,
.Ql X-Original-To
and
.Ql Status
fields.
The new
.Ql To
field contains the bounce address(es) plus any addresses in the old
.Ql To
field minus the user's local address and any on the alternates list.
(See the
.Ic alternates
command.)
.It Ic chdir
.Pq Ic c
Changes the user's working directory to that specified, if given.
If no directory is given, then changes to the user's login directory.
.It Ic copy
.Pq Ic co
The
.Ic copy
command does the same thing that
.Ic save
does, except that it does not mark the messages it
is used on for deletion when you quit.
.It Ic deldups
Delete duplicate messages based on their
.Ql Message-Id
field, keeping the first one in the current sort order.
This can be useful with replies to a mailing list that are also CCed
to a subscriber.
(The same thing can also be accomplished with the threading and
tagging commands.)
.It Ic delete
.Pq Ic d
Takes a list of messages as an argument and marks them all as deleted.
Deleted messages will not be saved in
.Ar mbox ,
nor will they be available for most other commands.
.It Ic detach
Takes a message list followed by a target directory as arguments,
decodes each MIME part in the message list, and saves it in the target
directory.
If the message list is empty, use the current message.
If the directory is not specified, use the directory specified by
.Ar mime-detach-dir
variable and, if that is empty, default to the directory
.Nm
was started in.
For each MIME part in the message list, the filename is displayed for
confirmation or changes.
If an empty name is entered, the part is skipped.
If the filename already exists, the user will be prompted before
overwriting it.
(See the
.Ar mime-detach-batch
and
.Ar mime-detach-overwrite
variables to change this behavior.)
Only MIME parts with an associated filename in the
.Ql Content-Type
or
.Ql Content-Disposition
fields are decoded.
(See
.Ic Detach
to detach all parts.)
The MIME extension hooks and character set conversion are ignored.
.It Ic dp
(also
.Ic dt )
Deletes the current message and prints the next message.
If there is no next message,
.Nm
says
.Dq Li "at EOF" .
.It Ic down
Go down one level in the thread.
If given a message number, it descends the thread below that message,
otherwise it descends from the current message (dot).
.It Ic edit
.Pq Ic e
Takes a list of messages and points the text editor at each one in
turn.
On return from the editor, the message is read back in.
.It Ic else
Switch the command execution condition set by the previous
.Ic if ,
.Ic ifdef ,
or
.Ic ifndef
command.
.It Ic endif
Terminate an
.Ic if ,
.Ic ifdef ,
or
.Ic ifndef
command.
.It Ic exit
.Po Ic ex
or
.Ic x
.Pc
Effects an immediate return to the Shell without
modifying the user's system mailbox, his
.Ar mbox
file, or his edit file in
.Fl f .
.It Ic expandaddr
If unset (the default), recipient addresses must be names of local mailboxes
or Internet mail addresses.
If set, shell expansion of existing mailbox names will be performed.
.It Ic expose
Expose the thread structure so all messages appear in header listings.
(See
.Ar hide
for the inverse.)

doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 7.0 to NetBSD 8.0:
	libc: Update to tzcode2014f. [christos 20140815]
	gmake:	Update to gmake-3.81 in external/gpl2/gmake and remove 
		gnu/dist/gmake (3.80) [christos 20140818]
	module(7): Create MODULAR_DEFAULT_AUTOLOAD config flag to control
		whether module autoloading is on by default.
		[jnemeth 20140824]
	kernel: Option SYMTAB_SPACE has been replaced with COPY_SYMTAB.
		The related changes to src/sys/conf require a rebuild of
		config(1). [joerg 20140824]
	zoneinfo: Import tzdata2014g. [apb 20140831]
	wm(4): Add internal SERDES support for 82575 and newer controllers.
		[msaitoh 20140903]
	or1k: Add new port.  [matt 20140903]
	awindma(4): Add support for Allwinner A10/A20 DMA controller.
		[jmcneill 20140906]
	awinac(4): Add support for Allwinner A10/A20 audio codec.
		[jmcneill 20140906]
	awinrtc(4): Add support for Allwinner A10/A20 real-time clock.
		[jmcneill 20140907]
	awinmmc(4): Use DMA for MMC transfers. [jmcneill 20140908]
	axp20x(4): Add AXP20x PMU driver. [jmcneill 20140909]
	rtsol(8): Removed in favour of dhcpcd. [roy 20140911]
	rtsold(8): Removed in favour of dhcpcd. [roy 20140911]
	dhcpcd(8): Import dhcpcd-6.4.5. [roy 20140918]
	arm: Add support for i.MX6 SoC. [ryo 20140925]
	gpt(8): Completed overhaul, including adding follwing subcommands:
		resize, set, unset, backup, restore, and resizedisk.
		[jnemeth 20140926]
	dhcpcd(8): Import dhcpcd-6.4.7. [roy 20140927]
	mpt(4): added bio(4) support. [jmcneill 20140927]
	gpt(8): added type subcommand to modify partition type.
		[jnemeth 20140927]
	zoneinfo: Import tzdata2014h. [apb 20141004]
	dhcpcd(8): Import dhcpcd-6.5.0. [roy 20141006]
	libc: Update to tzcode2014h. [christos 20141007]
	file(1): Upgraded to 5.20. [christos 20141010]
	arm: Add support for Allwinner A31 SoC. [jmcneill 20141010]
	awiniic(4): Add support for Allwinner A31 SoC. [jmcneill 20141012]
	awinp2wi(4): Add support for Allwinner A31 Push-Pull Two Wire
		Interface. [jmcneill 20141012]
	axp22x(4): Add AXP22x PMU driver. [jmcneill 20141012]
	awinrtc(4): Add support for Allwinner A31 real-time clock.
		[jmcneill 20141012]
	awindma(4): Add support for Allwinner A31 DMA controller.
		[jmcneill 20141013]
	awinac(4): Add support for Allwinner A31 audio codec.
		[jmcneill 20141013]
	motg(4): Add support for Allwinner A10/A20 SoC. [jmcneill 20141015]
	wpa: Import wpa_supplicant and hostapd 2.3. [christos 20141016]
	dhcpcd(8): Import dhcpcd-6.5.1. [roy 20141018]
	OpenSSH: Imported 6.7. [christos 20141018]
	resolvconf(8): Import openresolv-3.6.0 [roy 20141020]
	gcc: Support -fsanitize=address [christos 20141022]
	zoneinfo: Import tzdata2014i. [apb 20141023]
	libc: Update to tzcode2014i. [christos 20141023]
	openpam(3): update to 20140912 (ourouparia) [christos 20141024]
	pppd(8): updated to version 2.4.7. [christos 20141025]
	acpi(4): Updated ACPICA to 20140926. [christos 20141025]
	dhcpcd(8): Import dhcpcd-6.6.0 [roy 20141029]
	resolvconf(8): Import openresolv-3.6.1 [roy 20141029]
	rtsx(4): Add support for Realtek RTS5227/RTL8402/RTL8411/RTL8411B.
		[nonaka 20141029]
	flex(1): Import flex-2.5.37 [christos 20141029]
	iwn(4): Add support for Intel Centrino Wireless-N
		100/105/130/135/2200/2230. From OpenBSD. [nonaka 20141030]
	awge(4): Add support for gigabit ethernet found on various
		Allwinner SOCs. [martin 20141101]
	dhcpcd(8): Import dhcpcd-6.6.1. [roy 20141107]
	awinhdmi(4): Add support for Allwinner A20/A31 HDMI controller.
		[jmcneill 20141110]
	awintcon(4): Add support for Allwinner A20/A31 TV/LCD controller.
		[jmcneill 20141110]
	awindebe(4): Add support for Allwinner A20/A31 Display engine
		backend (DE-BE). [jmcneill 20141110]
	awinhdmiaudio(4): Add support for Allwinner A20/A31 HDMI audio
		controller. [jmcneill 20141111]
	dhcpcd(8): Import dhcpcd-6.6.2. [roy 20141114]
	awinir(4): Add support for Allwinner A20/A31 IR receiver.
		[jmcneill 20141115]
	zoneinfo: Import tzdata2014j. [apb 20141116]
	libc: Update to tzcode2014j. [christos 20141117]
	libpcap: Import 1.6.2. [christos 20141119]
	tcpdump(8): Import 4.6.2. [christos 20141119]
	dhcpcd(8): Import dhcpcd-6.6.4. [roy 20141126]
	dhcpcd(8): Import dhcpcd-6.6.5. [roy 20141209]
	bind: Import version 9.10.1-P1. [christos 20141209]
	dhcpcd(8): Import dhcpcd-6.6.6. [roy 20141217]
	powerpc: Switch to GCC 4.8. [matt 20141218]
	dhcpcd(8): Import dhcpcd-6.6.7. [roy 20141219]
	ntp: Import ntp 4.2.8. [christos 20141219]
	midirecord(1): Add program to record MIDI files.  [mrg 20141229]


libexec/mail.local/mail.local.c
/*	$NetBSD$	*/

/*-
 * Copyright (c) 1990, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1990, 1993, 1994\
 The Regents of the University of California.  All rights reserved.");
#if 0
static char sccsid[] = "@(#)mail.local.c	8.22 (Berkeley) 6/21/95";
#else
__RCSID("$NetBSD$");
#endif
#endif /* not lint */

#include <sys/param.h>
#include <sys/stat.h>
#include <sys/socket.h>

#include <netinet/in.h>

#include <errno.h>
#include <fcntl.h>
#include <pwd.h>
#include <netdb.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <time.h>
#include <unistd.h>
#include <sysexits.h>


#include "pathnames.h"

static int	deliver(int, char *, int);
__dead static void	logerr(int, const char *, ...) __printflike(2, 3);
static void	logwarn(const char *, ...) __printflike(1, 2);
static void	notifybiff(char *);
static int	store(const char *);
__dead static void	usage(void);

int
main(int argc, char *argv[])
{
	struct passwd *pw;
	int ch, fd, eval, lockfile = 0;
	uid_t uid;
	const char *from;

	/* use a reasonable umask */
	(void) umask(0077);

	openlog("mail.local", LOG_PERROR, LOG_MAIL);

	from = NULL;
	while ((ch = getopt(argc, argv, "ldf:r:")) != -1)
		switch (ch) {
		case 'd':		/* backward compatible */
			break;
		case 'f':
		case 'r':		/* backward compatible */
			if (from)
				logerr(EX_USAGE, "multiple -f options");
			from = optarg;
			break;
		case 'l':
			lockfile++;
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (!*argv)
		usage();

	/*
	 * If from not specified, use the name from getlogin() if the
	 * uid matches, otherwise, use the name from the password file
	 * corresponding to the uid.
	 */
	uid = getuid();
	if (!from && (!(from = getlogin()) ||
	    !(pw = getpwnam(from)) || pw->pw_uid != uid))
		from = (pw = getpwuid(uid)) ? pw->pw_name : "???";

	fd = store(from);
	for (eval = EX_OK; *argv; ++argv) {
		int rval;

		rval = deliver(fd, *argv, lockfile);
		if (eval == EX_OK && rval != EX_OK)
			eval = rval;
	}
	exit (eval);
}

static int
store(const char *from)
{
	FILE *fp = NULL;	/* XXX gcc */
	time_t tval;
	int fd, eline;
	char *tn, line[2048];

	tn = strdup(_PATH_LOCTMP);
	if (!tn)
		logerr(EX_OSERR, "not enough core");
	if ((fd = mkstemp(tn)) == -1 || !(fp = fdopen(fd, "w+")))
		logerr(EX_OSERR, "unable to open temporary file");
	(void)unlink(tn);
	free(tn);

	(void)time(&tval);
	(void)fprintf(fp, "From %s %s", from, ctime(&tval));

	line[0] = '\0';
	for (eline = 1; fgets(line, sizeof(line), stdin);) {
		if (line[0] == '\n')
			eline = 1;
		else {
			if (eline && line[0] == 'F' && !memcmp(line, "From ", 5))
				(void)putc('>', fp);
			eline = 0;
		}
		(void)fprintf(fp, "%s", line);
		if (ferror(fp))
			break;
	}

	/* If message not newline terminated, need an extra. */
	if (!index(line, '\n'))
		(void)putc('\n', fp);
	/* Output a newline; note, empty messages are allowed. */
	(void)putc('\n', fp);

	(void)fflush(fp);
	if (ferror(fp))
		logerr(EX_OSERR, "temporary file write error");
	fd = dup(fd);
	(void)fclose(fp);
	return(fd);
}

static int
deliver(int fd, char *name, int lockfile)
{
	struct stat sb;
	struct passwd pwres, *pw;
	char pwbuf[1024];
	int created, mbfd, nr, nw, off, rval=EX_OK, lfd=-1;
	char biffmsg[100], buf[8*1024], path[MAXPATHLEN], lpath[MAXPATHLEN];
	off_t curoff;

	/*
	 * Disallow delivery to unknown names -- special mailboxes can be
	 * handled in the sendmail aliases file.
	 */
	if ((getpwnam_r(name, &pwres, pwbuf, sizeof(pwbuf), &pw)) != 0) {
		logwarn("unable to find user %s: %s", name, strerror(errno));
		return(EX_TEMPFAIL);
	}
	if (pw == NULL) {
		logwarn("unknown name: %s", name);
		return(EX_NOUSER);
	}

	(void)snprintf(path, sizeof path, "%s/%s", _PATH_MAILDIR, name);

	if (lockfile) {
		(void)snprintf(lpath, sizeof lpath, "%s/%s.lock",
		    _PATH_MAILDIR, name);

		if((lfd = open(lpath, O_CREAT|O_WRONLY|O_EXCL,
		    S_IRUSR|S_IWUSR)) < 0) {
			logwarn("%s: %s", lpath, strerror(errno));
			return(EX_OSERR);
		}
	}

	if (!(created = lstat(path, &sb)) &&
	    (sb.st_nlink != 1 || S_ISLNK(sb.st_mode))) {
		logwarn("%s: linked file", path);
		return(EX_OSERR);
	}
	
	if ((mbfd = open(path, O_APPEND|O_WRONLY|O_EXLOCK,
	    S_IRUSR|S_IWUSR)) < 0) {
		if ((mbfd = open(path, O_APPEND|O_CREAT|O_WRONLY|O_EXLOCK,
		    S_IRUSR|S_IWUSR)) < 0) {
			logwarn("%s: %s", path, strerror(errno));
			return(EX_OSERR);
		}
	}

	curoff = lseek(mbfd, 0, SEEK_END);
	(void)snprintf(biffmsg, sizeof biffmsg, "%s@%lld\n", name,
	    (long long)curoff);
	if (lseek(fd, 0, SEEK_SET) == (off_t)-1) {
		logwarn("temporary file: %s", strerror(errno));
		rval = EX_OSERR;
		goto bad;
	}

	while ((nr = read(fd, buf, sizeof(buf))) > 0)
		for (off = 0; off < nr;  off += nw)
			if ((nw = write(mbfd, buf + off, nr - off)) < 0) {
				logwarn("%s: %s", path, strerror(errno));
				goto trunc;
			}
	if (nr < 0) {
		logwarn("temporary file: %s", strerror(errno));
trunc:		(void)ftruncate(mbfd, curoff);
		rval = EX_OSERR;
	}

	/*
	 * Set the owner and group.  Historically, binmail repeated this at
	 * each mail delivery.  We no longer do this, assuming that if the
	 * ownership or permissions were changed there was a reason for doing
	 * so.
	 */
bad:
	if (lockfile) {
		if (lfd >= 0) {
			unlink(lpath);
			close(lfd);
		}
	}
	if (created) 
		(void)fchown(mbfd, pw->pw_uid, pw->pw_gid);

	(void)fsync(mbfd);		/* Don't wait for update. */
	(void)close(mbfd);		/* Implicit unlock. */

	if (rval == EX_OK)
		notifybiff(biffmsg);

	return rval;
}

void
notifybiff(char *msg)
{
	static struct sockaddr_in addr;
	static int f = -1;
	struct hostent *hp;
	struct servent *sp;
	int len;

	if (!addr.sin_family) {
		/* Be silent if biff service not available. */
		if (!(sp = getservbyname("biff", "udp")))
			return;
		if (!(hp = gethostbyname("localhost"))) {
			logwarn("localhost: %s", strerror(errno));
			return;
		}
		addr.sin_len = sizeof(struct sockaddr_in);
		addr.sin_family = hp->h_addrtype;
		addr.sin_port = sp->s_port;
		memcpy(&addr.sin_addr, hp->h_addr, hp->h_length);
	}
	if (f < 0 && (f = socket(AF_INET, SOCK_DGRAM, 0)) == -1) {
		logwarn("socket: %s", strerror(errno));
		return;
	}
	len = strlen(msg) + 1;
	if (sendto(f, msg, len, 0, (struct sockaddr *)&addr, sizeof(addr))
	    != len)
		logwarn("sendto biff: %s", strerror(errno));
}

static void
usage(void)
{
	logerr(EX_USAGE, "usage: mail.local [-l] [-f from] user ...");
}

static void
logerr(int status, const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	vsyslog(LOG_ERR, fmt, ap);
	va_end(ap);

	exit(status);
	/* NOTREACHED */
}

static void
logwarn(const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	vsyslog(LOG_ERR, fmt, ap);
	va_end(ap);
	return;
}
